### PR TITLE
Internal: Add Behat tests for "Add to calendar" feature

### PR DESCRIPTION
**Pull request by nechai**
## Problem
As a developer I want to have Behat tests for "Add to calendar" feature.

## Solution
- Add Behat tests for **iCal** calendar

## Issue tracker
- https://getopensocial.atlassian.net/browse/PROD-25673

## Theme issue tracker
n/a

## How to test
n/a

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] New features or changes to existing features are covered by tests, either unit (preferably) or behat
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Screenshots
n/a

## Release notes
- Add Behat tests for **iCal** calendar

## Change Record
n/a

## Translations
n/a

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

File: tests/behat/features/bootstrap/EventContext.php [hunk 1]
```diff
@@ -541,6 +541,31 @@ public function createAnonymousEventEnrollees(TableNode $eventAnonymousEnrollees
     }
   }
 
+  /**
+   * Enable the calendar button with a given calendar.
+   *
+   * @Given ":calendar" calendar is enabled
+   */
+  public function iDisableVerifiedImmediately(string $calendar) {
```
Comment (ribel) [inline tests/behat/features/bootstrap/EventContext.php:549]: @nechai I guess we need to rename this method ;)

**Review by ribel — COMMENTED** [review]
@nechai Please change method name in EventContext.
I also would like to ask @Kingdutch to take a look and review this test.

File: tests/behat/features/bootstrap/EventContext.php (hunk 1 above)
Comment (nechai) [reply tests/behat/features/bootstrap/EventContext.php]: Done 😅

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

File: tests/behat/features/bootstrap/FileContext.php [hunk 2]
```diff
@@ -63,4 +74,29 @@ private function getUrlWithGuzzle(array $cookies, string $url) : ResponseInterfa
...
+    $hostname = (isset($_SERVER['HTTPS']) ? 'https' : 'http') . '://' . $_SERVER['HTTP_HOST'];
+    if (strpos($url, $hostname) === FALSE) {
+      $url = $hostname . $url;
+    }
+
+    $cookies = $this->getSession()->getDriver()->getCookies();
+    $response = $this->getUrlWithGuzzle($cookies, $url);
+
+    return trim($response->getBody()->getContents());
+  }
```
Comment (Kingdutch) [inline tests/behat/features/bootstrap/FileContext.php:100]: In case we can't re-use the `matches` step but need the new `contains` step, then I suggest two changes to this function:

- Make the method `protected` since we know in Cablecar we can extend the contexts for SaaS specific test steps
- Don't `trim` inside `getDownloadedFileContent` but do it in the calling function. This helps if we ever need to test anything related to downloaded whitespace (like a YAML file).

File: tests/behat/features/bootstrap/EventContext.php [hunk 3]
```diff
@@ -541,6 +541,31 @@ public function createAnonymousEventEnrollees(TableNode $eventAnonymousEnrollees
     }
   }
 
+  /**
+   * Enable the calendar button with a given calendar.
+   *
+   * @Given ":calendar" calendar is enabled
```
Comment (Kingdutch) [inline tests/behat/features/bootstrap/EventContext.php:547]: I propose rephrasing this slightly to help in discovery and also disambiguate from any internal calendar function we might have.

```suggestion
   * @Given add to calendar is enabled for :calendar
```

I've dropped the quotes around `:calendar` because `"iCal"` still works but it can also just be `iCal` (so it makes the quotes optional)

**Review by Kingdutch — COMMENTED** [review]
I think it's great to add test coverage for iCal and the chosen test format works well here :D 

Are we planning to do the same for Google or does it work differently?

File: tests/behat/features/capabilities/event/add-to-calendar.feature [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    Given users:
+      | name               | mail             | status | timezone | roles    |
+      | regular_user       | some@example.com | 1      | UTC      |verified |
+    Given events with non-anonymous author:
+      | title               | body        | field_content_visibility | field_event_date    | field_event_date_end | langcode |
+      | Walking in the park | lorem ipsum | public                   | 2100-01-01T10:00:00 | 2100-01-01T12:00:00  | en       |
+
+    When I am logged in as "regular_user"
+    And I am viewing the event "Walking in the park"
+    And the file downloaded from "iCal calendar" should contain contents:
```
[the remark this answers is not in the document]
Comment (nechai) [reply tests/behat/features/capabilities/event/add-to-calendar.feature:22]: I didn't add testing or description because it contain link to event and this link contain event `id` which is dynamic data....

File: tests/behat/features/capabilities/event/add-to-calendar.feature [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    When I click "Enroll as guest"
+    And I wait for AJAX to finish
+    And I fill in the following:
+      | First name    | John         |
+      | Last name     | Doe          |
+      | Email address | john@doe.com |
+    And I press "Enroll in event" in the "Modal"
+    And I wait for AJAX to finish
+    Then I should see the link "iCal calendar"
+    And the file downloaded from "iCal calendar" should contain contents:
```
[the remark this answers is not in the document]
Comment (nechai) [reply tests/behat/features/capabilities/event/add-to-calendar.feature:65]: Removed

File: tests/behat/features/bootstrap/FileContext.php (hunk 2 above)
Comment (nechai) [reply tests/behat/features/bootstrap/FileContext.php:100]: Added

File: tests/behat/features/bootstrap/EventContext.php (hunk 3 above)
Comment (nechai) [reply tests/behat/features/bootstrap/EventContext.php:547]: Added

File: tests/behat/features/capabilities/event/add-to-calendar.feature (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kingdutch) [reply tests/behat/features/capabilities/event/add-to-calendar.feature]: Ah I see :D Hadn't realized that, yeah then we need a separate step :D

**Review by Kingdutch — APPROVED** [review]
Tests look good to me! Nice job :D

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Review by ribel — APPROVED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3425-m6o51bax1gpwlacdb5qhaqijoyzulus2.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/64884e73f4c4559361c722be

**Comment by ribel** [comment]
🍒 picked to 11.9.x
